File: advent_of_code/day1/main.py
from typing import List, Optional, Tuple
from advent_of_code.helpers.helpers import read_file_line_by_line


def check_sum(x: int, y: int) -> Optional[Tuple[int, int, int]]:
    if x+y == 2020:
        return x, y, x*y


def cast_to_int(x: str) -> int:
    return int(x)


def make_pairs(data: List[int]) -> List[List[int]]:
    output: List[int] = []
    # i and j are the pairs
    for i in data:
        for j in data:
            if i != j:
                temp_list: List[int] = sorted([i, j])
                if temp_list not in output:
                    output.append(temp_list)
    return output


def main() -> None:
    input_data: List[str] = read_file_line_by_line(filepath="advent_of_code/day1/input.txt")
    input_integers: List[int] = [cast_to_int(x=row) for row in input_data]
    number_pairs: List[List[int]] = make_pairs(data=input_integers)

    for pair in number_pairs:
        res: Optional[Tuple[int, int, int]] = check_sum(x=pair[0], y=pair[1])
        if res is not None:
            print(res)


if __name__ == "__main__":
    main()

File: advent_of_code/day1/main.go
package main

import (
	"errors"
	"fmt"
	"sort"
	"strconv"

	helpers "github.com/thesammy2010/advent-of-code/advent_of_code/helpers"
)

func checkSum(x int, y int) (int, int, int, error) {
	if x+y == 2020 {
		return x, y, x * y, nil
	}
	return 0, 0, 0, errors.New("x & y do not add up to 2020")
}

func castToInt(x string) int {
	y, err := strconv.ParseInt(x, 10, 64)
	if err != nil {
		fmt.Errorf("%s", err)
	}
	return int(y)
}

func makePairs(data []int) [][]int {
	list := make([][]int, 0)
	// i and j are the pairs
	for _, i := range data {
		for _, j := range data {
			if i != j {
				tempList := []int{i, j}
				sort.Ints(tempList)
				list = append(list, tempList)
			}
		}
	}
	return list
}

func main() {
	inputData := helpers.ReadFileLineByLine("input.txt")
	inputDataIntegers := []int{}
	for _, row := range inputData {
		inputDataIntegers = append(inputDataIntegers, castToInt(row))
	}
	numberPairs := makePairs(inputDataIntegers)
	for _, row := range numberPairs {
		a, b, c, err := checkSum(row[0], row[1])
		if err != nil {
			fmt.Println(a, b, c)
		}
	}
}
